### PR TITLE
[JENKINS-38552] BranchBuildStrategy to skip the initial build on the first branch indexing

### PR DESCRIPTION
**Pull request by fbelzunc**
This PR intends to fix [JENKINS-38552](https://issues.jenkins-ci.org/browse/JENKINS-38552)

This PR is dependent of [this PR on branch-api-plugin](https://github.com/jenkinsci/branch-api-plugin/pull/144)

The idea is to not build on the first branch indexing as in a real enterprise context this might produce dozens/hundreds of agents to start builds, which I don't see why they are necessary - from the little I saw on the code.

### CODE REVIEW

File: src/main/java/jenkins/branch/buildstrategies/basic/SkipInitialBuildOnFirstBranchIndexing.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
+package jenkins.branch.buildstrategies.basic;
```
Comment (rsandell) [inline src/main/java/jenkins/branch/buildstrategies/basic/SkipInitialBuildOnFirstBranchIndexing.java]: Missing license

File: src/main/resources/jenkins/branch/buildstrategies/basic/SkipInitialBuildOnFirstBranchIndexing/help.html [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<!--
+ ~ The MIT License
+ ~
+ ~ Copyright (c) 2018, CloudBees, Inc.
```
Comment (rsandell) [inline src/main/resources/jenkins/branch/buildstrategies/basic/SkipInitialBuildOnFirstBranchIndexing/help.html]: 2019
It's a new year, a new life, a new dawn...

File: src/test/java/jenkins/branch/buildstrategies/basic/SkipInitialBuildOnFirstBranchIndexingTest.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import jenkins.scm.impl.mock.MockSCMController;
+import jenkins.scm.impl.mock.MockSCMHead;
+import jenkins.scm.impl.mock.MockSCMRevision;
+import jenkins.scm.impl.mock.MockSCMSource;
+import org.junit.Test;
+
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.*;
+
+
```
Comment (rsandell) [inline src/test/java/jenkins/branch/buildstrategies/basic/SkipInitialBuildOnFirstBranchIndexingTest.java]: I would prefer something more integration testy that integrates/checks the changes in https://github.com/jenkinsci/branch-api-plugin/pull/144

File: src/test/java/jenkins/branch/buildstrategies/basic/SkipInitialBuildOnFirstBranchIndexingTest.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
+package jenkins.branch.buildstrategies.basic;
```
Comment (rsandell) [inline src/test/java/jenkins/branch/buildstrategies/basic/SkipInitialBuildOnFirstBranchIndexingTest.java]: license + copyright

**Comment by fbelzunc** [comment]
@rsandell I think I did all the changes, please, review when you have the chance.

File: src/test/java/jenkins/branch/buildstrategies/basic/SkipInitialBuildOnFirstBranchIndexingTest.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2018, CloudBees, Inc.
```
Comment (rsandell) [inline src/test/java/jenkins/branch/buildstrategies/basic/SkipInitialBuildOnFirstBranchIndexingTest.java]:
```suggestion
 * Copyright (c) 2019, CloudBees, Inc.
```

File: src/test/java/jenkins/branch/buildstrategies/basic/SkipInitialBuildOnFirstBranchIndexingTest.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    private void fire(MockSCMHeadEvent event) throws Exception {
+        long watermark = SCMEvents.getWatermark();
+        SCMHeadEvent.fireNow(event);
+        SCMEvents.awaitAll(watermark);
+        j.waitUntilNoActivity();
+    }
+
+}
```
Comment (conn) [inline src/test/java/jenkins/branch/buildstrategies/basic/SkipInitialBuildOnFirstBranchIndexingTest.java]: The end of the file should be targeted by a newline

**Comment by fbelzunc** [comment]
I just updated the PR with the requested changes

**Comment by conn** [comment]
Great work @fbelzunc ! You definitely scratch an itch.

File: src/main/java/jenkins/branch/buildstrategies/basic/AllBranchBuildStrategyImpl.java [hunk 7]
```diff
@@ -77,7 +78,8 @@ public boolean isAutomaticBuild(@NonNull SCMSource source, @NonNull SCMHead head
                 source,
                 head,
                 currRevision,
-                prevRevision
+                prevRevision,
+                taskListener
```
Comment (drdamour) [inline src/main/java/jenkins/branch/buildstrategies/basic/AllBranchBuildStrategyImpl.java]: this is a problem...this strategy wraps other strategies from previous versions that might not yet have implemented this new interface with the extra method...as such it'll cause abstract method problem.

Comment (drdamour) [reply src/main/java/jenkins/branch/buildstrategies/basic/AllBranchBuildStrategyImpl.java]: i think you'd need to add a TaskAwareBuildStrategy interface that has this new method and check for instanceof that.  maybe there's a more common approach in jenins, but that's how i've done it in java in the past.

Comment (fbelzunc) [reply src/main/java/jenkins/branch/buildstrategies/basic/AllBranchBuildStrategyImpl.java]: Hi @drdamour I am fixing this in #5 - once the build is successful... would you mind to give a try, please?

Comment (drdamour) [reply src/main/java/jenkins/branch/buildstrategies/basic/AllBranchBuildStrategyImpl.java]: #5 won't fix this..in fact it makes it even worse as you have further diverged from the existing interface!

Comment (drdamour) [reply src/main/java/jenkins/branch/buildstrategies/basic/AllBranchBuildStrategyImpl.java]: https://issues.jenkins-ci.org/browse/JENKINS-56984

**Pull request merged**